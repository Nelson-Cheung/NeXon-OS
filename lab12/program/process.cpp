#include "program_manager.h"
#include "../cstdlib.h"
#include "../interrupt.h"
#include "tss.h"

// 用户进程初始化，构建用户进程上下文环境
void startProcess(void *filename)
{
    PCB *pcb = sysProgramManager.running();
    ThreadInterruptStack *interruptStack = (ThreadInterruptStack *)((dword)pcb->stack + sizeof(ThreadStack));

    interruptStack->edi = 0;
    interruptStack->esi = 0;
    interruptStack->ebp = 0;
    interruptStack->esp_dummy = 0;
    interruptStack->ebx = 0;
    interruptStack->ecx = 0;
    interruptStack->eax = 0;
    interruptStack->gs = 0;

    interruptStack->fs = 0x3b;
    interruptStack->es = 0x3b;
    interruptStack->ss = 0x3b;
    interruptStack->ds = 0x3b;

    interruptStack->eip = (dword)filename;
    interruptStack->cs = 0x33;                                // 用户模式平坦模式
    interruptStack->eflags = (3 << 12) | (1 << 9) | (1 << 1); // IOPL, IF, MBS
    interruptStack->esp = (dword)specifyPaddrForVaddr(AddressPoolType::USER, USER_STACK_VADDR) + PAGE_SIZE;

    sys_start_process((dword)interruptStack);
}

// 激活线程或进程页目录表
void ProgramManager::activatePageDir(PCB *program)
{
    dword paddr = 0x100000; // 内核页目录表物理地址

    if (program->pageDir)
    {
        paddr = vaddr2paddr((dword)program->pageDir);
    }

    sys_update_cr3(paddr);
}

// 激活线程或进程页表
void ProgramManager::activatePageTab(PCB *program)
{
    activatePageDir(program);

    if (program->pageDir)
    {
        tss.updateEsp0(program);
    }
}

// 创建用户进程的页目录表
dword *ProgramManager::createPageDir()
{
    dword *vaddr = (dword *)allocatePages(AddressPoolType::KERNEL, 1);
    if (!vaddr)
    {
        //printf("can not create page from kernel\n");
        return nullptr;
    }

    // 复制内核目录项到虚拟地址的高1GB
    dword *src = (dword *)(0xfffff000 + 0x300 * 4);
    dword *dst = (dword *)((dword)vaddr + 0x300 * 4);
    for (int i = 0; i < 256; ++i)
    {
        dst[i] = src[i];
    }

    // 最后一项设置为用户进程页目录表物理地址
    vaddr[1023] = vaddr2paddr((dword)vaddr) | 0x7;
    return vaddr;
}

// 创建用户虚拟地址池
void ProgramManager::createUserVaddrPool(PCB *pcb)
{
    dword sourcesCount = (0xc0000000 - USER_VADDR_START) / PAGE_SIZE;
    dword length = (sourcesCount + 8 - 1) / 8;
    // 计算位图所占的页数
    dword pagesCount = (length + PAGE_SIZE - 1) / PAGE_SIZE;

    void *start = allocatePages(AddressPoolType::KERNEL, pagesCount);
    (pcb->userVaddr).setResources((byte *)start, sourcesCount);
    (pcb->userVaddr).setStartAddress(USER_VADDR_START);
}

// 创建用户进程
dword ProgramManager::executeProcess(void *filename, const char *name, dword priority)
{
    PCB *process = buildThreadPCB(nullptr, nullptr, name, priority);
    // 创建进程特定部分

    // 创建页目录表
    process->pageDir = createPageDir();
    //创建用户地址池
    createUserVaddrPool(process);

    ThreadStack *threadStack = (ThreadStack *)process->stack;
    threadStack->eip = (dword)startProcess;
    threadStack->arg = filename;

    process->memoryManager.initialize();

    bool interruptStatus = _interrupt_status();
    _disable_interrupt();
    allPrograms.push_back(&(process->tagInAllList));
    readyPrograms.push_back(&(process->tagInGeneralList));
    _set_interrupt(interruptStatus);

    return process->pid;
}

PCB *ProgramManager::findChildProcess(dword parentPid) {
    ThreadListItem *item = allPrograms.head.next;
    PCB *child, *ans;

    ans = nullptr;

    bool status = _interrupt_status();
    _disable_interrupt();
    while(item) {
        child = threadListItem2PCB(item);
        if(child->parentPid == parentPid) {
            ans = child;
            break;
        }
    }
    _set_interrupt(status);

    return ans;
}