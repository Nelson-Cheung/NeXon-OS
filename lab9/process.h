#ifndef PROCESS_H
#define PROCESS_H

#include "thread.h"

#define USER_STACK_VADDR (0xc0000000 - 0x1000)
#define USER_VADDR_START 0x8048000

// 用户进程初始化，构建用户进程上下文环境
void startProcess(void *filename);
// 激活线程或进程页目录表
void activatePageDir(PCB *thread);
// 激活线程或进程页表
void activatePageTab(PCB *thread);
// 创建用户进程的页目录表
dword *createPageDir();
// 创建用户虚拟地址池
void createUserVaddrPool(PCB *pcb);
// 创建用户进程
void executeProcess(void *filename, char *name);

#endif