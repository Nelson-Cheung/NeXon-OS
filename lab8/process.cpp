#include "process.h"
#include "memory.h"
#include "tss.h"
#include "cstdio.h"
#include "cstdlib.h"
#include "thread.h"

extern "C" void sys_start_process(dword esp);
extern "C" dword sys_update_cr3(dword address);

void startProcess(void *filename)
{
    PCB *pcb = (PCB *)_running_thread();
    ThreadInterruptStack *interruptStack = (ThreadInterruptStack *)((dword)pcb->stack + sizeof(ThreadStack));

    interruptStack->edi =
        interruptStack->esi =
            interruptStack->ebp =
                interruptStack->esp_dummy =
                    interruptStack->ebx =
                        interruptStack->ecx =
                            interruptStack->eax =
                                interruptStack->gs = 0;

    interruptStack->fs =
        interruptStack->es =
            interruptStack->ss =
                interruptStack->ds = 0x3b;

    interruptStack->eip = (dword)filename;
    interruptStack->cs = 0x33;                                // 用户模式平坦模式
    interruptStack->eflags = (3 << 12) | (1 << 9) | (1 << 1); // IOPL, IF, MBS
    interruptStack->esp = (dword)specifyPaddrForVaddr(AddressPoolType::USER, USER_STACK_VADDR) + PAGE_SIZE;

    sys_start_process((dword)interruptStack);
}

void activatePageDir(PCB *pcb)
{
    dword paddr = 0x100000; // 内核页目录表物理地址

    if (pcb->pageDir)
    {
        paddr = vaddr2paddr((dword)pcb->pageDir);
    }

    sys_update_cr3(paddr);
}

void activatePageTab(PCB *pcb)
{
    activatePageDir(pcb);

    if (pcb->pageDir)
    {
        tss.updateEsp0(pcb);
    }
}

dword *createPageDir()
{
    dword *vaddr = (dword *)allocatePages(AddressPoolType::KERNEL, 1);
    if (!vaddr)
    {
        //printf("can not create page from kernel\n");
        return nullptr;
    }

    // 复制内核目录项到虚拟地址的高1GB
    dword *src = (dword *)(0xfffff000 + 0x300 * 4);
    dword *dst = (dword *)((dword)vaddr + 0x300 * 4);
    for (int i = 0; i < 256; ++i)
    {
        dst[i] = src[i];
    }

    // 最后一项设置为用户进程页目录表物理地址
    vaddr[1023] = vaddr2paddr((dword)vaddr) | 0x7;
    return vaddr;
}

void createUserVaddrPool(PCB *pcb)
{
    dword sourcesCount = (0xc0000000 - USER_VADDR_START) / PAGE_SIZE;
    dword length = (sourcesCount + 8 - 1) / 8;
    // 计算位图所占的页数
    dword pagesCount = (length + PAGE_SIZE - 1) / PAGE_SIZE;

    void *start = allocatePages(AddressPoolType::KERNEL, pagesCount);
    (pcb->userVaddr).setResources((byte *)start, sourcesCount);
    (pcb->userVaddr).setStartAddress(USER_VADDR_START);
}
dword executeProcess(void *filename, const char *name, dword priority)
{
    PCB *process = (PCB *)allocatePages(AddressPoolType::KERNEL, 1);
    if (!process)
        return -1;

    memset((byte *)process, 0, PAGE_SIZE);

    process->status = ThreadStatus::READY;
    process->priority = priority;
    process->ticks = priority;
    process->ticksPassedBy = 0;
    process->pageDir = createPageDir();
    createUserVaddrPool(process);
    process->magic = 0x19870916;

    process->stack = (dword *)((dword)process + PAGE_SIZE - sizeof(ThreadInterruptStack) - sizeof(ThreadStack));

    ThreadStack *threadStack = (ThreadStack *)process->stack;
    threadStack->eip = kernelThread;
    threadStack->function = startProcess;
    threadStack->funcArg = filename;
    threadStack->ebp =
        threadStack->ebx =
            threadStack->esi =
                threadStack->edi = 0;

    bool interruptStatus = _interrupt_status();
    _disable_interrupt();
    _all_threads.push_back(&(process->tagInAllList));
    _ready_threads.push_back(&(process->tagInGeneralList));
    _set_interrupt(interruptStatus);

    return process->pid;
}