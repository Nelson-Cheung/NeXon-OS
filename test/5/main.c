#include "interface.h"
#include "process.h"
#define MAX_BUFLEN 30 
extern void loadUsrProgram(int seg, int addr, int len, int head, int cylinder, int sector);

void shell(){
    int usrprogNum=sizeof(Comlist) / sizeof(struct sector);
    enum commandEnum{  Help,Cls,Shutdown,Run,List};
    char userbuff[MAX_BUFLEN]={0};
    char command[MAX_BUFLEN]={0};
    char target[MAX_BUFLEN]={0};
    char *emptystr = "\0";
    char *commands[] = {"help","cls","shutdown","run","ls"};
    cls();
    printInfo();
    while (1)
    {
        prompt();
        readbuff(userbuff, MAX_BUFLEN);
        if(strcmp(userbuff,emptystr)==0){
            continue;
        }
        if(strcmp(userbuff, commands[Shutdown]) == 0) {
            shutdown();
        }
        else if(strcmp(userbuff, commands[Cls]) == 0) {
            cls();
        }
        else if(strcmp(userbuff, commands[Help]) == 0) {
            printInfo();
        }
        else if(strcmp(userbuff, commands[List]) == 0) {
            print(ComlistTitle);
            newline;
            for (int i = 0; i < usrprogNum; i++)
            {
                printDetail(Comlist[i]);
                newline;
            }
        }
        else {
            int valid=0;
            split(userbuff, command, target);
            if(strcmp(command, commands[Run]) == 0) {
                valid =target[0]!='\0'&& checknum(target,usrprogNum);
            }
            if(valid){
                for (int i = 0, len = strlen(target); i < len; i++)
                {
                    int pid = target[i] - '0'-1;
                    loadUsrProgram(Comlist[pid].seg,Comlist[pid].addr, Comlist[pid].len, Comlist[pid].head, Comlist[pid].cylinder, Comlist[pid].sector);
                    cls();
                }
            }
            else{
                char *error_msg = "Bad command.Please try other.Key in \"help\" for help\r\n";
                print(error_msg);
            }
            
        }
    }
    
}

