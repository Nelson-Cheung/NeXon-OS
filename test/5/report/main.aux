\relax 
\providecommand\hyper@newdestlabel[2]{}
\providecommand\HyperFirstAtBeginDocument{\AtBeginDocument}
\HyperFirstAtBeginDocument{\ifx\hyper@anchor\@undefined
\global\let\oldcontentsline\contentsline
\gdef\contentsline#1#2#3#4{\oldcontentsline{#1}{#2}{#3}}
\global\let\oldnewlabel\newlabel
\gdef\newlabel#1#2{\newlabelxx{#1}#2}
\gdef\newlabelxx#1#2#3#4#5#6{\oldnewlabel{#1}{{#2}{#3}}}
\AtEndDocument{\ifx\hyper@anchor\@undefined
\let\contentsline\oldcontentsline
\let\newlabel\oldnewlabel
\fi}
\fi}
\global\let\hyper@last\relax 
\gdef\HyperFirstAtBeginDocument#1{#1}
\providecommand*\HyPL@Entry[1]{}
\HyPL@Entry{0<</S/D>>}
\citation{ostextbook}
\@writefile{toc}{\contentsline {section}{\numberline {1}实验目的}{2}{section.1}\protected@file@percent }
\@writefile{toc}{\contentsline {section}{\numberline {2}实验要求}{2}{section.2}\protected@file@percent }
\@writefile{toc}{\contentsline {section}{\numberline {3}实验内容}{2}{section.3}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {3.1}编写用于中断处理的现场保护和现场恢复的汇编过程}{2}{subsection.3.1}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {3.2}重写和扩展实验三的的内核程序}{2}{subsection.3.2}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {3.3}实现部分基本输入输出库过程}{2}{subsection.3.3}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {3.4}执行调用自己设计的库过程的程序}{2}{subsection.3.4}\protected@file@percent }
\@writefile{toc}{\contentsline {section}{\numberline {4}实验原理}{2}{section.4}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {4.1}中断}{2}{subsection.4.1}\protected@file@percent }
\citation{intprocess}
\citation{intprocess}
\@writefile{lot}{\contentsline {table}{\numberline {1}{\ignorespaces 中断的分类\relax }}{3}{table.caption.3}\protected@file@percent }
\providecommand*\caption@xref[2]{\@setref\relax\@undefined{#1}}
\newlabel{tab:class-int}{{1}{3}{中断的分类\relax }{table.caption.3}{}}
\@writefile{toc}{\contentsline {paragraph}{中断的分类}{3}{section*.2}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {4.2}中断术语}{3}{subsection.4.2}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {4.3}中断处理的过程}{3}{subsection.4.3}\protected@file@percent }
\@writefile{toc}{\contentsline {section}{\numberline {5}实验过程}{4}{section.5}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {5.1}编写中断向量操作程序}{5}{subsection.5.1}\protected@file@percent }
\@writefile{toc}{\contentsline {subsubsection}{\numberline {5.1.1}移动中断向量宏}{5}{subsubsection.5.1.1}\protected@file@percent }
\newlabel{MOVE VECTOR}{{1}{5}{MOVE\_VECTOR}{lstlisting.1}{}}
\@writefile{lol}{\contentsline {lstlisting}{\numberline {1}MOVE\_VECTOR}{5}{lstlisting.1}\protected@file@percent }
\@writefile{toc}{\contentsline {subsubsection}{\numberline {5.1.2}中断向量写入宏}{5}{subsubsection.5.1.2}\protected@file@percent }
\newlabel{VECTOR IN}{{2}{5}{VECTOR\_IN}{lstlisting.2}{}}
\@writefile{lol}{\contentsline {lstlisting}{\numberline {2}VECTOR\_IN}{5}{lstlisting.2}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {5.2}时钟中断的响应处理程序}{6}{subsection.5.2}\protected@file@percent }
\newlabel{TimerTest}{{3}{6}{Timer(test)}{lstlisting.3}{}}
\@writefile{lol}{\contentsline {lstlisting}{\numberline {3}Timer(test)}{6}{lstlisting.3}\protected@file@percent }
\@writefile{toc}{\contentsline {paragraph}{主模块}{7}{section*.4}\protected@file@percent }
\@writefile{toc}{\contentsline {paragraph}{风火轮模块}{7}{section*.5}\protected@file@percent }
\@writefile{lof}{\contentsline {figure}{\numberline {1}{\ignorespaces 风火轮程序运行过程截图\relax }}{7}{figure.caption.6}\protected@file@percent }
\newlabel{fig:mixtest}{{1}{7}{风火轮程序运行过程截图\relax }{figure.caption.6}{}}
\@writefile{toc}{\contentsline {subsection}{\numberline {5.3}键盘中断的响应处理程序}{7}{subsection.5.3}\protected@file@percent }
\newlabel{ouch.asm}{{4}{7}{ouch.asm}{lstlisting.4}{}}
\@writefile{lol}{\contentsline {lstlisting}{\numberline {4}ouch.asm}{7}{lstlisting.4}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {5.4}修改内核和用户程序}{8}{subsection.5.4}\protected@file@percent }
\@writefile{toc}{\contentsline {subsubsection}{\numberline {5.4.1}修改内核}{8}{subsubsection.5.4.1}\protected@file@percent }
\newlabel{kernelChange}{{5}{8}{内核程序更改}{lstlisting.5}{}}
\@writefile{lol}{\contentsline {lstlisting}{\numberline {5}内核程序更改}{8}{lstlisting.5}\protected@file@percent }
\@writefile{toc}{\contentsline {subsubsection}{\numberline {5.4.2}修改用户程序}{9}{subsubsection.5.4.2}\protected@file@percent }
\newlabel{userChange}{{6}{9}{用户程序更改}{lstlisting.6}{}}
\@writefile{lol}{\contentsline {lstlisting}{\numberline {6}用户程序更改}{9}{lstlisting.6}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {5.5}扩展内核程序-实时显示时间}{9}{subsection.5.5}\protected@file@percent }
\@writefile{toc}{\contentsline {subsubsection}{\numberline {5.5.1}从BIOS获得日期时间的信息}{9}{subsubsection.5.5.1}\protected@file@percent }
\@writefile{lot}{\contentsline {table}{\numberline {2}{\ignorespaces 日期时间枚举表\relax }}{9}{table.caption.7}\protected@file@percent }
\newlabel{tab:enum-date}{{2}{9}{日期时间枚举表\relax }{table.caption.7}{}}
\@writefile{toc}{\contentsline {subsubsection}{\numberline {5.5.2}加工和显示获得的信息}{10}{subsubsection.5.5.2}\protected@file@percent }
\newlabel{split}{{7}{10}{split}{lstlisting.7}{}}
\@writefile{lol}{\contentsline {lstlisting}{\numberline {7}split}{10}{lstlisting.7}\protected@file@percent }
\@writefile{toc}{\contentsline {subsubsection}{\numberline {5.5.3}修改中断程序}{11}{subsubsection.5.5.3}\protected@file@percent }
\@writefile{toc}{\contentsline {subsubsection}{\numberline {5.5.4}在合适的位置加载中断向量}{11}{subsubsection.5.5.4}\protected@file@percent }
\@writefile{toc}{\contentsline {paragraph}{修改内核程序}{11}{section*.8}\protected@file@percent }
\@writefile{toc}{\contentsline {paragraph}{修改加载用户程序的函数}{11}{section*.9}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {5.6}软盘扇区安排}{11}{subsection.5.6}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {5.7}镜像文件生成}{11}{subsection.5.7}\protected@file@percent }
\@writefile{lot}{\contentsline {table}{\numberline {3}{\ignorespaces 软盘扇区安排\relax }}{11}{table.caption.10}\protected@file@percent }
\newlabel{tab:sectortable}{{3}{11}{软盘扇区安排\relax }{table.caption.10}{}}
\@writefile{toc}{\contentsline {section}{\numberline {6}程序使用说明}{12}{section.6}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {6.1}实验环境}{12}{subsection.6.1}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {6.2}编译方法}{12}{subsection.6.2}\protected@file@percent }
\@writefile{toc}{\contentsline {subsubsection}{\numberline {6.2.1}系统要求}{12}{subsubsection.6.2.1}\protected@file@percent }
\@writefile{toc}{\contentsline {subsubsection}{\numberline {6.2.2}编译过程与参数}{12}{subsubsection.6.2.2}\protected@file@percent }
\newlabel{code:generateos}{{8}{12}{生成镜像}{lstlisting.8}{}}
\@writefile{lol}{\contentsline {lstlisting}{\numberline {8}生成镜像}{12}{lstlisting.8}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {6.3}运行与演示}{12}{subsection.6.3}\protected@file@percent }
\@writefile{toc}{\contentsline {subsubsection}{\numberline {6.3.1}风火轮}{12}{subsubsection.6.3.1}\protected@file@percent }
\@writefile{lof}{\contentsline {figure}{\numberline {2}{\ignorespaces 带风火轮的shell\relax }}{13}{figure.caption.11}\protected@file@percent }
\newlabel{fig:sheel}{{2}{13}{带风火轮的shell\relax }{figure.caption.11}{}}
\@writefile{toc}{\contentsline {subsubsection}{\numberline {6.3.2}带时间显示和风火轮}{13}{subsubsection.6.3.2}\protected@file@percent }
\@writefile{lof}{\contentsline {figure}{\numberline {3}{\ignorespaces 带时间显示和风火轮的shell进入界面\relax }}{13}{figure.caption.12}\protected@file@percent }
\newlabel{fig:wheelentr}{{3}{13}{带时间显示和风火轮的shell进入界面\relax }{figure.caption.12}{}}
\@writefile{lof}{\contentsline {figure}{\numberline {4}{\ignorespaces 带时间显示和风火轮的shell\relax }}{14}{figure.caption.13}\protected@file@percent }
\newlabel{fig:wheelShell}{{4}{14}{带时间显示和风火轮的shell\relax }{figure.caption.13}{}}
\@writefile{toc}{\contentsline {subsubsection}{\numberline {6.3.3}带时间显示和风火轮}{14}{subsubsection.6.3.3}\protected@file@percent }
\@writefile{lof}{\contentsline {figure}{\numberline {5}{\ignorespaces 按下任意键后的"Ouch!Ouch!"\relax }}{14}{figure.caption.14}\protected@file@percent }
\newlabel{fig:timeDisp}{{5}{14}{按下任意键后的"Ouch!Ouch!"\relax }{figure.caption.14}{}}
\citation{makefile}
\citation{lamport94}
\citation{cite}
\@writefile{toc}{\contentsline {section}{\numberline {7}总结与讨论}{15}{section.7}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {7.1}特色,不足与改进}{15}{subsection.7.1}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {7.2}收获}{15}{subsection.7.2}\protected@file@percent }
\@writefile{toc}{\contentsline {subsection}{\numberline {7.3}遇到的问题和感想}{15}{subsection.7.3}\protected@file@percent }
\@writefile{toc}{\contentsline {paragraph}{}{15}{section*.15}\protected@file@percent }
\@writefile{toc}{\contentsline {paragraph}{}{15}{section*.16}\protected@file@percent }
\@writefile{toc}{\contentsline {paragraph}{}{15}{section*.17}\protected@file@percent }
\@writefile{toc}{\contentsline {paragraph}{}{15}{section*.18}\protected@file@percent }
\bibcite{lamport94}{1}
\bibcite{cite}{2}
\bibcite{intprocess}{3}
\bibcite{ostextbook}{4}
\bibcite{makefile}{5}
\@writefile{lof}{\contentsline {figure}{\numberline {6}{\ignorespaces 旧用户程序的bug：按下键盘后显示的字符会变色\relax }}{16}{figure.caption.19}\protected@file@percent }
\newlabel{fig:olduserbug}{{6}{16}{旧用户程序的bug：按下键盘后显示的字符会变色\relax }{figure.caption.19}{}}
\@writefile{toc}{\contentsline {section}{\numberline {8}附录}{16}{section.8}\protected@file@percent }
\newlabel{utilsForDisp}{{9}{16}{加工和显示日期时间}{lstlisting.9}{}}
\@writefile{lol}{\contentsline {lstlisting}{\numberline {9}加工和显示日期时间}{16}{lstlisting.9}\protected@file@percent }
