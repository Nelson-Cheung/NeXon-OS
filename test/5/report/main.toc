\contentsline {section}{\numberline {1}实验目的}{2}{section.1}%
\contentsline {section}{\numberline {2}实验要求}{2}{section.2}%
\contentsline {section}{\numberline {3}实验内容}{2}{section.3}%
\contentsline {subsection}{\numberline {3.1}编写用于中断处理的现场保护和现场恢复的汇编过程}{2}{subsection.3.1}%
\contentsline {subsection}{\numberline {3.2}重写和扩展实验三的的内核程序}{2}{subsection.3.2}%
\contentsline {subsection}{\numberline {3.3}实现部分基本输入输出库过程}{2}{subsection.3.3}%
\contentsline {subsection}{\numberline {3.4}执行调用自己设计的库过程的程序}{2}{subsection.3.4}%
\contentsline {section}{\numberline {4}实验原理}{2}{section.4}%
\contentsline {subsection}{\numberline {4.1}中断}{2}{subsection.4.1}%
\contentsline {paragraph}{中断的分类}{3}{section*.2}%
\contentsline {subsection}{\numberline {4.2}中断术语}{3}{subsection.4.2}%
\contentsline {subsection}{\numberline {4.3}中断处理的过程}{3}{subsection.4.3}%
\contentsline {section}{\numberline {5}实验过程}{4}{section.5}%
\contentsline {subsection}{\numberline {5.1}编写中断向量操作程序}{5}{subsection.5.1}%
\contentsline {subsubsection}{\numberline {5.1.1}移动中断向量宏}{5}{subsubsection.5.1.1}%
\contentsline {subsubsection}{\numberline {5.1.2}中断向量写入宏}{5}{subsubsection.5.1.2}%
\contentsline {subsection}{\numberline {5.2}时钟中断的响应处理程序}{6}{subsection.5.2}%
\contentsline {paragraph}{主模块}{7}{section*.4}%
\contentsline {paragraph}{风火轮模块}{7}{section*.5}%
\contentsline {subsection}{\numberline {5.3}键盘中断的响应处理程序}{7}{subsection.5.3}%
\contentsline {subsection}{\numberline {5.4}修改内核和用户程序}{8}{subsection.5.4}%
\contentsline {subsubsection}{\numberline {5.4.1}修改内核}{8}{subsubsection.5.4.1}%
\contentsline {subsubsection}{\numberline {5.4.2}修改用户程序}{9}{subsubsection.5.4.2}%
\contentsline {subsection}{\numberline {5.5}扩展内核程序-实时显示时间}{9}{subsection.5.5}%
\contentsline {subsubsection}{\numberline {5.5.1}从BIOS获得日期时间的信息}{9}{subsubsection.5.5.1}%
\contentsline {subsubsection}{\numberline {5.5.2}加工和显示获得的信息}{10}{subsubsection.5.5.2}%
\contentsline {subsubsection}{\numberline {5.5.3}修改中断程序}{11}{subsubsection.5.5.3}%
\contentsline {subsubsection}{\numberline {5.5.4}在合适的位置加载中断向量}{11}{subsubsection.5.5.4}%
\contentsline {paragraph}{修改内核程序}{11}{section*.8}%
\contentsline {paragraph}{修改加载用户程序的函数}{11}{section*.9}%
\contentsline {subsection}{\numberline {5.6}软盘扇区安排}{11}{subsection.5.6}%
\contentsline {subsection}{\numberline {5.7}镜像文件生成}{11}{subsection.5.7}%
\contentsline {section}{\numberline {6}程序使用说明}{12}{section.6}%
\contentsline {subsection}{\numberline {6.1}实验环境}{12}{subsection.6.1}%
\contentsline {subsection}{\numberline {6.2}编译方法}{12}{subsection.6.2}%
\contentsline {subsubsection}{\numberline {6.2.1}系统要求}{12}{subsubsection.6.2.1}%
\contentsline {subsubsection}{\numberline {6.2.2}编译过程与参数}{12}{subsubsection.6.2.2}%
\contentsline {subsection}{\numberline {6.3}运行与演示}{12}{subsection.6.3}%
\contentsline {subsubsection}{\numberline {6.3.1}风火轮}{12}{subsubsection.6.3.1}%
\contentsline {subsubsection}{\numberline {6.3.2}带时间显示和风火轮}{13}{subsubsection.6.3.2}%
\contentsline {subsubsection}{\numberline {6.3.3}带时间显示和风火轮}{14}{subsubsection.6.3.3}%
\contentsline {section}{\numberline {7}总结与讨论}{15}{section.7}%
\contentsline {subsection}{\numberline {7.1}特色,不足与改进}{15}{subsection.7.1}%
\contentsline {subsection}{\numberline {7.2}收获}{15}{subsection.7.2}%
\contentsline {subsection}{\numberline {7.3}遇到的问题和感想}{15}{subsection.7.3}%
\contentsline {paragraph}{}{15}{section*.15}%
\contentsline {paragraph}{}{15}{section*.16}%
\contentsline {paragraph}{}{15}{section*.17}%
\contentsline {paragraph}{}{15}{section*.18}%
\contentsline {section}{\numberline {8}附录}{16}{section.8}%
