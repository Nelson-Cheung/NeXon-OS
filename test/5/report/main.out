\BOOKMARK [1][-]{section.1}{实验目的}{}% 1
\BOOKMARK [1][-]{section.2}{实验要求}{}% 2
\BOOKMARK [1][-]{section.3}{实验内容}{}% 3
\BOOKMARK [2][-]{subsection.3.1}{编写用于中断处理的现场保护和现场恢复的汇编过程}{section.3}% 4
\BOOKMARK [2][-]{subsection.3.2}{重写和扩展实验三的的内核程序}{section.3}% 5
\BOOKMARK [2][-]{subsection.3.3}{实现部分基本输入输出库过程}{section.3}% 6
\BOOKMARK [2][-]{subsection.3.4}{执行调用自己设计的库过程的程序}{section.3}% 7
\BOOKMARK [1][-]{section.4}{实验原理}{}% 8
\BOOKMARK [2][-]{subsection.4.1}{中断}{section.4}% 9
\BOOKMARK [2][-]{subsection.4.2}{中断术语}{section.4}% 10
\BOOKMARK [2][-]{subsection.4.3}{中断处理的过程}{section.4}% 11
\BOOKMARK [1][-]{section.5}{实验过程}{}% 12
\BOOKMARK [2][-]{subsection.5.1}{编写中断向量操作程序}{section.5}% 13
\BOOKMARK [2][-]{subsection.5.2}{时钟中断的响应处理程序}{section.5}% 14
\BOOKMARK [2][-]{subsection.5.3}{键盘中断的响应处理程序}{section.5}% 15
\BOOKMARK [2][-]{subsection.5.4}{修改内核和用户程序}{section.5}% 16
\BOOKMARK [2][-]{subsection.5.5}{扩展内核程序-实时显示时间}{section.5}% 17
\BOOKMARK [2][-]{subsection.5.6}{软盘扇区安排}{section.5}% 18
\BOOKMARK [2][-]{subsection.5.7}{镜像文件生成}{section.5}% 19
\BOOKMARK [1][-]{section.6}{程序使用说明}{}% 20
\BOOKMARK [2][-]{subsection.6.1}{实验环境}{section.6}% 21
\BOOKMARK [2][-]{subsection.6.2}{编译方法}{section.6}% 22
\BOOKMARK [2][-]{subsection.6.3}{运行与演示}{section.6}% 23
\BOOKMARK [1][-]{section.7}{总结与讨论}{}% 24
\BOOKMARK [2][-]{subsection.7.1}{特色,不足与改进}{section.7}% 25
\BOOKMARK [2][-]{subsection.7.2}{收获}{section.7}% 26
\BOOKMARK [2][-]{subsection.7.3}{遇到的问题和感想}{section.7}% 27
\BOOKMARK [1][-]{section.8}{附录}{}% 28
